k2/csrc/fsa_equivalent.cc

```diff
@@ -62,8 +90,63 @@ bool IsRandEquivalent(const Fsa &a, const Fsa &b, std::size_t npath /*=100*/) {
   return true;
 }
 
-bool RandomPath(const Fsa &a, Fsa *b,
-                std::vector<int32_t> *state_map /*=nullptr*/) {
+template <class FbWeightType>
+bool IsRandEquivalent(const Fsa &a, const float *a_weights, const Fsa &b,
+                      const float *b_weights, std::size_t npath = 100) {
+  Fsa connected_a, connected_b, valid_a, valid_b;
+  std::vector<int32_t> connected_a_arc_map, connected_b_arc_map,
+      valid_a_arc_map, valid_b_arc_map;
+  Connect(a, &connected_a, &connected_a_arc_map);
+  Connect(b, &connected_b, &connected_b_arc_map);
+  ArcSort(connected_a, &valid_a, &valid_a_arc_map);  // required by `intersect`
+  ArcSort(connected_b, &valid_b, &valid_b_arc_map);
+  if (IsEmpty(valid_a) && IsEmpty(valid_b)) return true;
+  if (IsEmpty(valid_a) || IsEmpty(valid_b)) return false;
+
+  // Get arc weights
+  std::vector<float> valid_a_weights(valid_a.arcs.size());
+  std::vector<float> valid_b_weights(valid_b.arcs.size());
+  GetArcWeights(a_weights, connected_a_arc_map, valid_a_arc_map,
+                &valid_a_weights);
+  GetArcWeights(b_weights, connected_b_arc_map, valid_b_arc_map,
+                &valid_b_weights);
+
+  // Check that arc labels are compatible.
+  std::unordered_set<int32_t> labels_a, labels_b, labels_difference;
+  for (const auto &arc : valid_a.arcs) labels_a.insert(arc.label);
+  for (const auto &arc : valid_b.arcs) labels_b.insert(arc.label);
+  SetDifference(labels_a, labels_b, &labels_difference);
+  if (labels_difference.size() >= 2 ||
+      (labels_difference.size() == 1 &&
+       (*(labels_difference.begin()) != kEpsilon)))
+    return false;
+
+  std::random_device rd;
+  std::mt19937 gen(rd());
+  std::bernoulli_distribution coin(0.5);
+  for (auto i = 0; i != npath; ++i) {
+    const auto &fsa = coin(gen) ? valid_a : valid_b;
+    Fsa path, valid_path;
+    RandomPathWithoutEpsilonArc(fsa, &path);  // path is already connected
+    ArcSort(path, &valid_path);
+
+    Fsa a_compose_path, b_compose_path;
+    std::vector<float> a_compose_weights, b_compose_weights;
+    Intersect(valid_a, valid_a_weights.data(), path, &a_compose_path,
+              &a_compose_weights);
+    Intersect(valid_b, valid_b_weights.data(), path, &b_compose_path,
+              &b_compose_weights);
+    double sum_a = ShortestDistance<FbWeightType>(a_compose_path,
```

Comment:
@danpovey, Seems we cannot suppose `a_compose_path` is top-sorted and need to implement the version ShortestDistance with relax things, right?

Reply:
I can't find those commits on github so can't se the whole thing, but if that's an issue, it would be possible to top-sort those things separately, no?  I'm open to having a version of ShortestDistance that doesn't require top-sorting though, or to make the whole thing not require top-sorting if it would be as fast.   (But I do slightly worry about whether that would be as easy to do on GPU).

Reply:
This is a new PR for checking equivalence for weighted FSA which will be used to check equivalence of `determinize/RmEpsilon/Interset..`. There's now no any implementation of `ShortestDistance` for any version in our repo.

In the above scenarios (determinze/RmEpsilon/Interset..), if both the input and output of these algorithms are top-sorted FSAs, then we can just only implement a version of `ShortestDistance` for top-sorted FSA, if not, we need to implement a version for non-sorted FSA. (if two FSAs are top-sorted, their intersection will also be top-sorted according to our implemeation). 

BTW, the current implementation of `RandEquivalent` does not require top-sorting in all steps except in `ShortestDistance`, that's why I wonder if we should provide a general version, or just TopSorted before calling `ShorestDistance` (may fail if there is cycle and then we will fail to the equivalence)

Reply:
How about adding a boolean flag that's true if the input is top-sorted.

Eventually we should add the notion of properties.  But there are some other changes that I want to make first, and I want to wait a bit on that.

Reply:
I'm OK with add a boolean flag.

And what's your changes? As this PR is not finished (including tests), I'm working on the PR of `RmEpsilon` now. I can do this PR first if you think it has a higher priority.